src/meta-srv/src/procedure/region_migration/downgrade_leader_region.rs

```diff
@@ -13,23 +13,506 @@
+            &format!("Datanode-{}@{}", leader.id, leader.addr),
+            common_time::util::current_time_millis(),
+            downgrade_instruction,
+        )
+        .with_context(|_| error::SerializeToJsonSnafu {
+            input: downgrade_instruction.to_string(),
+        })?;
+
+        let ch = Channel::Datanode(leader.id);
+        let receiver = ctx
+            .mailbox
+            .send(&ch, msg, DOWNGRADE_LEADER_REGION_TIMEOUT)
+            .await?;
+
+        match receiver.await? {
+            Ok(msg) => {
+                let reply = HeartbeatMailbox::json_reply(&msg)?;
+                let InstructionReply::DowngradeRegion(DowngradeRegionReply {
+                    last_entry_id,
+                    exists,
+                    error,
+                }) = reply
+                else {
+                    return error::UnexpectedInstructionReplySnafu {
+                        mailbox_message: msg.to_string(),
+                        reason: "expect downgrade region reply",
+                    }
+                    .fail();
+                };
+
+                if error.is_some() {
+                    return error::RetryLaterSnafu {
+                        reason: format!(
+                            "Failed to downgrade the region {} on Datanode {:?}, error: {:?}",
+                            region_id, leader, error
+                        ),
+                    }
+                    .fail();
+                }
+
+                if !exists {
+                    warn!(
+                        "Trying to downgrade the region {} on Datanode {}, but region doesn't exist!",
+                        region_id, leader
+                    );
+                }
+
+                if let Some(last_entry_id) = last_entry_id {
+                    ctx.volatile_ctx.set_last_entry_id(last_entry_id);
+                }
+
+                Ok(())
+            }
+            Err(error::Error::MailboxTimeout { .. }) => {
+                let reason = format!(
+                    "Mailbox received timeout for downgrade leader region {region_id} on Datanode {:?}", 
+                    leader,
+                );
+                error::RetryLaterSnafu { reason }.fail()
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    /// Downgrades a leader region.
+    ///
+    /// Fast path:
+    /// - Waits for the reply of downgrade instruction.
+    ///
+    /// Slow path:
+    /// - Waits for the lease of the leader region expired.
+    async fn downgrade_region_with_retry(&self, ctx: &mut Context) {
+        let instruction = self.build_downgrade_region_instruction(ctx);
+
+        let mut retry = 0;
+
+        while retry < self.optimistic_retry {
+            if let Err(err) = self.downgrade_region(ctx, &instruction).await {
+                if err.is_retryable() {
+                    warn!("Failed to downgrade region, error: {err:?}, retry later.");
```

Comment:
The log information is inaccurate, because it won't retry if `retry + 1` = `optimistic_retry`